client-spark/spark3/src/main/java/org/apache/spark/shuffle/SpillableSizeTrackingMap.java

```diff
@@ -70,12 +71,23 @@ public SpillableSizeTrackingMap(TaskMemoryManager taskMemoryManager, Long spillT
         resetSamples();
     }
 
-    private void doSpill(Map currentMap) {
+    private void doSpillAndReleaseMemory(Map currentMap) {
+        long total = Math.max(0, grantedMemorySize - initialRequireMemorySize);
+        long released = 0;
         for (Iterator<Map.Entry<K, V>> it = currentMap.entrySet().iterator(); it.hasNext();) {
             Map.Entry<K, V> entry = it.next();
             it.remove();
+            long kvSize = SizeEstimator.estimate(entry);
```

Comment:
Method `SizeEstimator.estimate` have poor performance. Maybe we need some performance tests here. 
Maybe we can also release a inaccurate memory,  we remove a entry, we release (total * 1.0) / (currentMap.size()) memory. but it may cause oom.
It's a trade off. We need some tests.